### PR TITLE
Refactor lookaheads

### PR DESCRIPTION
**Pull request by maximedenes**
We move @ppedrot 's nice DSL from Ltac2 to `Pcoq` to make it available to other users, and we take advantage of it to refactor some existing code.

### CODE REVIEW

File: plugins/ltac/g_tactic.mlg [hunk 1]
```diff
@@ -107,11 +86,10 @@ let check_for_coloneq =
...
 
 let lookup_at_as_comma =
-  Pcoq.Entry.of_parser "lookup_at_as_comma"
-    (fun _ strm ->
-      match stream_nth 0 strm with
-        | KEYWORD (","|"at"|"as") -> ()
-        | _ -> err ())
+  let open Pcoq.Lookahead in
+  entry_of_lookahead "lookup_at_as_comma" begin
+    lk_kw "," <+> lk_kw "at" <+> lk_kw "as"
```
Comment (ppedrot) [inline plugins/ltac/g_tactic.mlg]: Use `lk_kws` instead.

File: parsing/pcoq.mli [hunk 2]
```diff
@@ -34,6 +34,20 @@ module Entry : sig
   val name : 'a t -> string
 end
 
+module Lookahead : sig
+  type lookahead
```
Comment (ppedrot) [inline parsing/pcoq.mli]: By convention, this should be called `t`.

File: parsing/pcoq.mli [hunk 3]
```diff
@@ -40,11 +40,13 @@ module Lookahead : sig
   val contiguous : (int -> Loc.t) -> int -> int -> bool
```
Comment (ppedrot) [inline parsing/pcoq.mli]: Isn't this dead code now?

**Review by herbelin — APPROVED** [review]
That's nice!

A suggestion: `lk_int` -> `lk_num` by consistency with the name of the token.

Note also that in #11688, I needed a generalization of `test_no_space` if ever you want to anticipate it as soon as now.

File: parsing/pcoq.mli (hunk 2 above)
Comment (maximedenes) [reply parsing/pcoq.mli]: Right, fixed.

File: parsing/pcoq.mli (hunk 3 above)
Comment (maximedenes) [reply parsing/pcoq.mli]: Indeed, I removed it from the interface.

File: plugins/ltac/g_tactic.mlg (hunk 1 above)
Comment (maximedenes) [reply plugins/ltac/g_tactic.mlg]: Right, I introduced it after and forgot to come back to this example. Fixed.

**Comment by maximedenes** [comment]
> A suggestion: `lk_int` -> `lk_num` by consistency with the name of the token.

Well, this is probably a question for @ppedrot, since he picked the name, but if you look at the implementation of `lk_int`, it is more restrictive than `NUMERAL`, which seems to contain information for non-integer numbers. If that's the case, all matches on this token I found were incorrect, and are fixed once ported to the DSL.

**Comment by ppedrot** [comment]
I didn't write the restricted `lk_int` version, it was actually introduced by @proux01 in 0d8a4ec.

**Comment by proux01** [comment]
> > A suggestion: `lk_int` -> `lk_num` by consistency with the name of the token.
> 
> Well, this is probably a question for @ppedrot, since he picked the name, but if you look at the implementation of `lk_int`, it is more restrictive than `NUMERAL`, which seems to contain information for non-integer numbers.

Right (c.f. `interp/numTok.mli`)

> If that's the case, all matches on this token I found were incorrect, and are fixed once ported to the DSL.

Do you mean, non-integer numbers where accepted in places only integers should have been? I did my best to avoid that when implementing #8764 . I'm really sorry if I missed something.

**Comment by proux01** [comment]
So I doubled check all instances of `NUMERAL` and found a few others looking suspicious/wrong:
* https://github.com/coq/coq/blob/c160fc0da9bef60c4ee469cc2c35afd83fc16243/parsing/g_constr.mlg#L275
* https://github.com/coq/coq/blob/c160fc0da9bef60c4ee469cc2c35afd83fc16243/parsing/g_constr.mlg#L387
this last two could be replaced by `bigint` (c.f., g_prim.mlg) if we want to only parse integers
* https://github.com/coq/coq/blob/c160fc0da9bef60c4ee469cc2c35afd83fc16243/plugins/ssr/ssrparser.mlg#L241 (4 instances)

**Comment by gares** [comment]
Awesome!

**Comment by herbelin** [comment]
>  If that's the case, all matches on this token I found were incorrect, and are fixed once ported to the DSL.

Indeed!

Then, may I suggest that it is renamed into `lk_nat` so that it is clearer that it is a positive integer. (I'm going in parallel to do similar changes in forthcoming PRs to clarify the respective role of positive and negative numbers.)

**Comment by maximedenes** [comment]
> may I suggest that it is renamed into `lk_nat`

Sure, why not. I'll update the patch.

**Comment by maximedenes** [comment]
> Sure, why not. I'll update the patch.

Done.

File: parsing/g_constr.mlg [hunk 4]
```diff
@@ -54,48 +53,25 @@ let err () = raise Stream.Failure
...
-    (fun _ strm ->
-      match stream_nth 0 strm with
-        | KEYWORD "{" ->
-            (match stream_nth 1 strm with
-              | IDENT ("wf"|"struct"|"measure") -> ()
-              | _ -> err ())
-        | _ -> err ())
+  let open Pcoq.Lookahead in
+  to_entry "check_fixannot" begin
+    lk_kw "{" >> (lk_kw "wf" <+> lk_kw "struct" <+> lk_kw "measure")
```
Comment (ppedrot) [inline parsing/g_constr.mlg]: That's another use of `lk_kws`. It's a bit strange, I think I remember it used that function when I first reviewed?

Comment (maximedenes) [reply parsing/g_constr.mlg]: Fixed, thanks.

**Comment by maximedenes** [comment]
@ppedrot This should be ready, can you self assign?

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Last round of comments before merge.

**Pull request merged**